### PR TITLE
[SQL][SPARK-5453] Use property 'mapreduce.input.pathFilter.class' to set a custom filter class for input files

### PR DESCRIPTION
**Pull request by saucam**
This PR adds support for using a custom filter class for input files for queries. We can re-use the existing property in hive-site.xml for this.

### CODE REVIEW

**Comment by marmbrus** [comment]
Sorry for the delay reviewing this.  At a high level this seems like a reasonable thing to support.  I'd suggest that we consolidate the logic for constructing the `Option[PathFilter]` into a single location in `SQLConf` instead of duplicating it.  We should also document that we support his property in the programming guide and audit the code base to make sure we aren't missing any places where we list files.

**Comment by saucam** [comment]
Thanks for the suggestions @marmbrus , I have refactored PathFilter creation in SQLContext. Covered more instances of listStatus. please review.

**Comment by marmbrus** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala [hunk 1]
```diff
@@ -201,6 +201,22 @@ class SparkHadoopUtil extends Logging {
...
     if (baseStatus.isDir) recurse(basePath) else Array(baseStatus)
   }
+
+  /**
+   * Get [[FileStatus]] objects for all leaf children (files) under the given base path. If the
+   * given path points to a file, return a single-element collection containing [[FileStatus]] of
+   * that file. Uses pathFilter to filter out certain filestatuses
+   */
+  def listLeafStatusesFiltered(fs: FileSystem, basePath: Path,
+    filter: Option[PathFilter]): Seq[FileStatus] = {
```
Comment (marmbrus) [inline core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala]: Indent as follows: https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide
```scala
def listLeafStatusesFiltered(
    fs: FileSystem,
    basePath: Path,
    filter: Option[PathFilter]): Seq[FileStatus] = {
  ...
```

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 2]
```diff
@@ -162,6 +166,26 @@ class SQLContext(@transient val sparkContext: SparkContext)
...
+    @Override
+    def accept(file: Path): Boolean = {
+      return true
+    }
+  }
+
+  /**
+   * Return PathFilter to be used for filtering input files using custom filter class
+   */
+  protected[sql] def getPathFilter(): Option[PathFilter] = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Why is this an option if it always returns `Some`?

**Comment by saucam** [comment]
please retest

**Comment by marmbrus** [comment]
Thanks for addressing the comments.  As I look at this closer I wonder if we aren't approaching this in the wrong way.  If this is a hadoop option that FileInputFormat supports, why aren't users just setting it in the hadoop configuration?  (i.e. `sc.hadoopConfiguration.set("mapreduce.input.pathFilter", ...)`).  As it stands now, it is sprinkling new logic all over that is reading from some random hadoop config.  This seems very likely to lead to errors in the future.  If we aren't properly propagating the hadoop config that is another thing that should be fixed separately.

Unless there is something that I am missing, I suggest that we close this issue and take the hadoop configuration approach.

If you do think this is the right approach, we can discuss further, but I have some feeback on the PR in general:
 - The indentation is still off in several places.  Please look at the coding style guide.
 - There is a lot of duplicated logic (some of which is wrong).  It would be better to isolate these things, so if we need to fix it in the future we can.  For example I see this patter a lot:

```scala
val filterClassName = sqlContext.getConf(NewFileInputFormat.PATHFILTER_CLASS, "")
  if (filterClassName.nonEmpty) {
    val filterClazz = Class.forName(filterClassName).asInstanceOf[Class[_ <: PathFilter]]
```

However `Class.forName` is probably not using a classloader that can see user code.

**Comment by saucam** [comment]
hey @marmbrus , this property is needed because : 
1. The 'hijacked' parquet read path would not use mapreduce property while reading schema/footers, 

see  refresh method in ParquetRelation2.scala

2. HiveTableScan would need the pathfilter while creating RDD 

see makeRDDForPartitionedTable in TableReader.scala

These are not using the mapreduce.input.pathFilter (even if its is set by the user) property and hence the extra code.

**Pull request closed without merging**